bee-protocol/src/worker/peer/message_handler.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::message::Header;
+
+use bee_network::Address;
+
+use futures::{
+    channel::{mpsc, oneshot},
+    future, select,
+    stream::{self, StreamExt},
+};
+
+use log::debug;
+
+type EventRecv = stream::Fuse<mpsc::Receiver<Vec<u8>>>;
+type ShutdownRecv = future::Fuse<oneshot::Receiver<()>>;
+
+/// The read state of the message handler.
+///
+/// This type is used by `MessageHandler` to decide what should be read next when handling an
+/// event.
+enum ReadState {
+    /// `MessageHandler` should read a header.
+    Header,
+    /// `MessageHandler` should read a payload based on a header.
+    Payload(Header),
+}
+
+/// A message handler.
+///
+/// It takes care of processing events into messages that can be processed by the workers.
+pub(super) struct MessageHandler {
+    events: EventHandler,
+    state: ReadState,
+    /// The address of the peer. This field is only here for logging purposes.
+    address: Address,
+}
+
+impl MessageHandler {
+    /// Create a new message handler from an event receiver, a shutdown receiver and the peer's
+    /// address.
+    pub(super) fn new(receiver: EventRecv, shutdown: ShutdownRecv, address: Address) -> Self {
+        Self {
+            events: EventHandler::new(receiver, shutdown),
+            // The handler should read a header first.
+            state: ReadState::Header,
+            address,
+        }
+    }
+    /// Fetch the header and payload of a message.
+    pub(super) async fn fetch_message<'a>(&'a mut self) -> Option<(Header, &'a [u8])> {
+        // loop until we can return the header and payload
+        loop {
+            match &self.state {
+                // Read a header.
+                ReadState::Header => {
+                    // We need 3 bytes to read a header.
+                    let bytes = self.events.fetch_bytes(3).await?;
```

Comment:
Maybe use the constant `HEADER_SIZE` instead of 3

Reply:
sure!